appliance/mariadb/cmd/flynn-mariadb-api/main.go

```diff
@@ -119,9 +137,173 @@ func (a *API) dropDatabase(ctx context.Context, w http.ResponseWriter, req *http
+	if a.scaledUp {
+		return nil
+	}
+
+	app := "mariadb"
+	logger := a.logger().New("fn", "scaleUp")
+
+	// Connect to controller.
+	logger.Info("connecting to controller")
+	client, err := controller.NewClient("", os.Getenv("CONTROLLER_KEY"))
+	if err != nil {
+		logger.Error("controller client error", "err", err)
+		return err
+	}
+
+	// Retrieve mariadb release.
+	logger.Info("retrieving app release", "app", app)
+	release, err := client.GetAppRelease(app)
+	if err == controller.ErrNotFound {
+		logger.Error("release not found", "app", app)
+		return errors.New("mariadb release not found")
+	} else if err != nil {
+		logger.Error("get release error", "app", app, "err", err)
+		return err
+	}
+
+	// Retrieve current formation.
+	logger.Info("retrieving formation", "app", app, "release_id", release.ID)
+	formation, err := client.GetFormation(app, release.ID)
+	if err == controller.ErrNotFound {
+		logger.Error("formation not found", "app", app, "release_id", release.ID)
+		return errors.New("mariadb formation not found")
+	} else if err != nil {
+		logger.Error("formation error", "app", app, "release_id", release.ID, "err", err)
+		return err
+	}
+
+	// If mariadb is running then exit.
+	if formation.Processes["mariadb"] > 0 {
+		logger.Info("database is running, scaling not necessary")
+		return nil
+	}
+
+	// Copy processes and increase database processes.
+	processes := make(map[string]int, len(formation.Processes))
+	for k, v := range formation.Processes {
+		processes[k] = v
+	}
+
+	if os.Getenv("SINGLETON") == "true" {
+		processes["mariadb"] = 1
+	} else {
+		processes["mariadb"] = 3
+	}
+
+	// Otherwise begin scaling the database processes up.
+	logger.Info("watching discoverd events", "app", app, "release_id", release.ID)
+	events := make(chan *discoverd.Event)
+	if _, err := discoverd.DefaultClient.Service("mariadb").Watch(events); err != nil {
+		logger.Error("watch discoverd events error", "app", app, "err", err)
+		return err
+	}
+
+	// Update formation.
+	logger.Info("updating formation", "app", app, "release_id", release.ID)
+	formation.Processes = processes
+	if err := client.PutFormation(formation); err != nil {
+		logger.Error("put formation error", "app", app, "release_id", release.ID, "err", err)
+		return err
+	}
+
+	// Wait for discoverd event from service.
+	logger.Info("waiting for scale up", "app", app, "release_id", release.ID)
+	if err := a.waitForUpEvent(events); err != nil {
+		return err
+	}
+
+	logger.Info("pinging new processes")
+	ticker := time.NewTicker(500 * time.Millisecond)
+	defer ticker.Stop()
+	timer := time.NewTimer(5 * time.Minute)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-timer.C:
+			return errors.New("timed out while starting mariadb cluster")
+		case <-ticker.C:
+		}
+
+		// Attempt to connect to database.
+		if err := func() error {
+			db, err := a.connect()
+			if err != nil {
+				return err
+			}
+			defer db.Close()
+
+			if _, err := db.Exec("SELECT 1"); err != nil {
```

Comment:
I think the `WaitForReadWrite` method in the sirenia client is more appropriate.